packages/salesforcedx-vscode-soql/src/queryDataView/queryDataFileService.ts

```diff
@@ -46,33 +47,47 @@ export class QueryDataFileService {
     }
   }
 
-  public save(): string {
+  public async save(): Promise<string> {
+    let queryDataSelectedPath = '';
     const fileContent = this.dataProvider.getFileContent(
       this.queryData.records
     );
-    const savedFileName = this.dataProvider.getFileName();
-    const queryDataFilePath = path.join(
-      this.getResultsDirectoryPath(),
-      savedFileName
+    const defaultFileName = this.dataProvider.getFileName();
+    /* queryDataDefaultFilePath will be used as the default options in the save dialog
+        fileName: The name of the soqlFile viewed in the builder
+        path: the same directory as the .soql file text doc.
+    note: directory must exist to show up in save dialog.
+    */
+    const queryDataDefaultFilePath = path.join(
+      path.parse(this.document.uri.path).dir,
+      defaultFileName
     );
 
-    this.createResultsDirectoryIfDoesNotExist();
-    // Save query results to disk
-    fs.writeFileSync(queryDataFilePath, fileContent);
-    this.showSaveSuccessMessage(savedFileName);
-    this.showFileInExplorer(queryDataFilePath);
+    const saveDialogOptions = {
+      defaultUri: vscode.Uri.file(queryDataDefaultFilePath)
+    };
 
-    return queryDataFilePath;
-  }
+    queryDataSelectedPath = await vscode.window
+      .showSaveDialog({
+        ...saveDialogOptions
+      })
+      .then((fileInfo: vscode.Uri | undefined) => {
+        if (fileInfo) {
+          queryDataSelectedPath = fileInfo.path;
+          // Save query results to disk
+          fs.writeFileSync(fileInfo.path, fileContent);
+          // Only reveal saved file if its inside current workspace
+          if (fileInfo.path.startsWith(getRootWorkspacePath())) {
+            this.showFileInExplorer(fileInfo.path);
+          }
+          this.showSaveSuccessMessage(path.basename(fileInfo.path));
 
-  private getResultsDirectoryPath() {
-    return path.join(getRootWorkspacePath(), QUERY_RESULTS_DIR_PATH);
-  }
+          return fileInfo.path;
+        }
+        return '';
+      });
```

Comment:
We can get rid of the promise methods and just use async/await syntax here:

```typescript
let resultsPath = '';

const fileInfo = await vscode.window.showSaveDialog(saveDialogOptions);

if (fileInfo) {
  resultsPath = fileInfo.path;
  fs.writeFileSync(resultsPath, fileContent);
  if (resultsPath.startsWith(getRootWorkspacePath())) {
    this.showFileInExplorer(resultsPath);
  }
  this.showSaveSuccessMessage(path.basename(resultsPath));
}

return resultsPath;
```

Reply:
ooo yeah, of course, good idea =)

Reply:
Done in the `fix: clean up code with async/await` commit